modules/cart/commerce_cart.module

```diff
@@ -103,6 +104,20 @@ function commerce_cart_commerce_order_delete(OrderInterface $order) {
 }
 
 /**
+ * Implements hook_ENTITY_TYPE_update().
+ */
+function commerce_cart_commerce_order_presave(OrderInterface $order) {
+  if (!$order->isNew() && $order->cart->value == TRUE) {
+    $place_transition = $order->getState()->getWorkflow()->getTransition('place');
+    $from_states = $place_transition->getFromStates();
+    $to_state = $place_transition->getToState()->getId();
+    if (in_array($order->original->state->value, array_keys($from_states)) && $order->state->value == $to_state) {
+      \Drupal::service('commerce_cart.cart_manager')->finalizeCart($order);
```

Comment:
We agreed to move the finalizeCart method from CartManager to CartProvider, that way we don't need to inject the CartProvider into the CartManager.